### PR TITLE
Feature/allow one non prefixed locale with shared host

### PR DESCRIPTION
**Pull request by manoj-krishnan**
This feature will allow sets of locales to share a hostname, with one locale from each set having the ability to exist at the root of the hostname. This is allowed for cases where a website may have a locale configuration which places the home page of each locale as follows,

'en' - example.com, (English - Generic)
'en-UK' - example.com/uk, (English - UK)
'en-DE' - example.com/de, (English - Germany)
'de' - example.de, (German - Generic)
'de-CH' - example.de/ch (German - Switzerland).

This configuration doesn't require a prefix for the 'en' and the 'de' locales. However, it requires a prefix for all other locales sharing the hostnames of these two root-level locales.

An extra check has been added to the `composeOptions` method to ensure that the hostname/prefix configuration satisfies the above conditions. This check will throw an error in cases where the configuration results in a conflict between two locales.

### CODE REVIEW

File: lib/middleware.js [hunk 1]
```diff
@@ -87,12 +87,25 @@ module.exports = function(self, options) {
...
-            return self.prefixes[candidate] === prefix;
+            return !_.has(self.prefixes, candidate);
           });
+        } else {
+          matches = req.url.match(/^\/[^/]+/);
+          if (matches) {
+            prefix = matches[0];
+            candidates = _.filter(candidates, function (candidate) {
+              return self.prefixes[candidate] === prefix;
+            });
```
Comment (boutell) [inline lib/middleware.js]: I don't see how this is going to do the right thing when there is no prefix (en or de in your example) and the URL isn't the root, but rather, let's say, /articles?

File: test/test2.js [hunk 2]
```diff
@@ -121,13 +138,34 @@ describe('Workflow Subdomains and Prefixes', function() {
...
+  it('can find a jointly-determined locale via middleware - case 2', function (done) {
+    tryMiddleware('http://example.es/co', function (req) {
+      assert(req.locale === 'es-CO');
+      done();
+    });
+  });
+
+  it('can detect a root-level locale via middleware - case 1', function(done) {
+    tryMiddleware('http://example.com', function(req) {
+      assert(req.locale === 'us-en');
```
Comment (boutell) [inline test/test2.js]: Looks like you don't have a test for a non-root page in this situation.

File: lib/middleware.js (hunk 1 above)
Comment (manoj-krishnan) [reply lib/middleware.js]: Ah, thanks for the catch! This is an embarrassing miss from my end. Guess I got carried away by the home page edge cases since I was primarily having trouble with those pages. I pushed a fix just now!

File: test/test2.js (hunk 2 above)
Comment (manoj-krishnan) [reply test/test2.js]: As part of my update, I've added a couple of tests to cover the cases you've mentioned.

**Comment by manoj-krishnan** [comment]
@boutell Tom, I've also added a few extra tests (apart from the ones for the case you mentioned) to cover the cases of locales under different hostnames, but sharing the same prefix. Wanted to make sure we have that explicitly covered in tests, even though the tests for hostname-based-filtering should guarantee us that we don't face issues with this case. Let me know if you think these tests are superfluous and need to be removed.

**Comment by boutell** [comment]
Thanks very much Manoj!
